crest/Assets/Crest/Crest/Scripts/Underwater/UnderwaterRenderer.Mask.cs

```diff
@@ -124,54 +179,79 @@ internal static void SetUpMaskTextures(CommandBuffer buffer, RenderTextureDescri
-            }
+            Helpers.CreateRenderTargetTexture(ref _boundaryFrontFaceRT, ref _boundaryFrontFaceTarget, descriptor);
+            _boundaryFrontFaceRT.name = "_CrestBoundaryFrontFaceTexture";
 
-            if (Instance._depthRT != null)
+            if (_mode == Mode.Geometry3D || _mode == Mode.GeometryVolume)
             {
-                Instance._depthRT.Release();
-                Instance._depthRT = null;
+                Helpers.CreateRenderTargetTexture(ref _boundaryBackFaceRT, ref _boundaryBackFaceTarget, descriptor);
+                _boundaryBackFaceRT.name = "_CrestBoundaryBackFaceTexture";
             }
         }
 
+        internal void CleanUpBoundaryTextures()
+        {
+            Helpers.DestroyRenderTargetTexture(ref _boundaryFrontFaceRT);
+            Helpers.DestroyRenderTargetTexture(ref _boundaryBackFaceRT);
+        }
+
+        internal void CleanUpMaskTextures()
+        {
+            Helpers.DestroyRenderTargetTexture(ref _maskRT);
+            Helpers.DestroyRenderTargetTexture(ref _depthRT);
+        }
+
+        internal void SetUpBoundary(Material maskMaterial)
+        {
+            Helpers.SetGlobalKeyword(k_KeywordBoundary2D, _mode == Mode.Geometry2D);
+            Helpers.SetGlobalKeyword(k_KeywordBoundaryHasBackFace, _mode == Mode.Geometry3D || _mode == Mode.GeometryVolume);
+            maskMaterial.SetKeyword(k_KeywordBoundary, _mode != Mode.FullScreen);
+            maskMaterial.SetInt("_StencilRef", IsStencilBufferRequired ? k_StencilValueBoundary : 0);
+        }
+
         void OnPreRenderOceanMask()
         {
+            _oceanMaskCommandBuffer.Clear();
+
             RenderTextureDescriptor descriptor = XRHelpers.GetRenderTextureDescriptor(_camera);
 
-            _oceanMaskCommandBuffer.Clear();
-            // Must call after clear or temporaries will be cleared.
-            SetUpMaskTextures(_oceanMaskCommandBuffer, descriptor);
-            _oceanMaskCommandBuffer.SetRenderTarget(_maskTarget, _depthTarget);
-            // Horizon pass makes clearing color redundant. If the horizon enabled ZWrite then we could skip clearing.
-            _oceanMaskCommandBuffer.ClearRenderTarget(true, false, Color.black);
-            _oceanMaskCommandBuffer.SetGlobalTexture(sp_CrestOceanMaskTexture, _maskTarget);
-            _oceanMaskCommandBuffer.SetGlobalTexture(sp_CrestOceanMaskDepthTexture, _depthTarget);
+            // Keywords and other things.
+            SetUpBoundary(_oceanMaskMaterial.material);
+            SetUpMaskTextures(descriptor);
 
-            SetInverseViewProjectionMatrix(_oceanMaskMaterial.material);
+            // Populate water boundary before mask so we can use the stencil.
+            if (_mode != Mode.FullScreen)
+            {
+                SetUpBoundaryTextures(descriptor);
+                PopulateBoundary(_oceanMaskCommandBuffer, _boundaryFrontFaceTarget, _boundaryBackFaceTarget);
+                if (IsStencilBufferRequired)
```

Comment:
Just realised I should be using the stencil buffer here regardless as it rejects a lot of the mask processing.

Reply:
Done.